Add docker support

clifinger opened the pull request:
This pull request install the Docker configuration for Bedrock project, as well as the build and extract script.

## Changes made

### docker-compose.yml
- Added configuration for a 3-node Bedrock cluster
- Set up networks and volumes for each node
- Defined node-specific environment variables

### Dockerfile
- Implemented a two-stage build to optimize final image size
- Used Ubuntu 22.04 as the base for the build stage
- Installed and configured apt-fast to speed up downloads
- Installed Clang 18 and necessary dependencies
- Configured ccache to improve compilation times
- Compiled the Bedrock project and ran tests
- Used phusion/baseimage:noble-1.0.0 for the final image
- Configured the Bedrock service to run with proper permissions

### build_and_extract.sh
- Added a script to build the Docker image and extract the Bedrock binary
- The script creates a temporary container, copies the binary, and places it in the output directory

## Motivation
These changes aim to improve our Docker build process, facilitate Bedrock cluster deployment, and optimize the size and performance of our Docker images.

## How to test
1. Ensure you have Docker and Docker Compose installed on your machine
2. Clone this branch and cd into `docker` folder
3. Run `docker-compose up --build` to build and start the Bedrock cluster
4. Verify that all three nodes start correctly and communicate with each other
5. Run the `build_and_extract.sh` script and check that the Bedrock binary is correctly extracted

Please feel free to provide any feedback or suggestions for improvement!

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+# Returns 1 if version doesn't match
+function bashVersionCheck()
+{
+  [ -z $BASH_VERSIO ] && return 1
```
```suggestion
  [ -z $BASH_VERSION ] && return 1
```

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+  [ -z $BASH_VERSIO ] && return 1
+
+  # If it's set, check the version
+  case $BASH_VERSIO in
```
```suggestion
  case $BASH_VERSION in
```

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+  esac
+}
+
+if [ $(bashVersionCheck) ]; then
```
We prefer to use `[[` over `[`

```suggestion
if [[ $(bashVersionCheck) ]]; then
```

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+NO_ARG_PARAMS=("VERBOSE" "QUIET" "CLEAN")
+ONE_ARG_PARAMS=("SERVER_HOST" "NODE_NAME" "PEER_LIST" "PRIORITY" "PLUGINS" "CACHE_SIZE" "WORKER_THREADS" "QUERY_LOG" "MAX_JOURNAL_SIZE" "SYNCHRONOUS")
+
+function toLowerCase()
+{
```
Our style is to keep the opening brace on the same line

```suggestion
function toLowerCase() {
```

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+  fi
+done
+
+exec /sbin/setuser bedrock /usr/local/bin/bedrock $PARAMS >> /var/log/bedrock.log 2>&1
```
Add a trailing new line please.

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+  fi
+done
+
+for i in "${ONE_ARG_PARAMS[@]}"
+do
+  :
+  if [ -v "$i" ]; then
```
```suggestion
for i in "${ONE_ARG_PARAMS[@]}" ; do
  if [[ -v "$i" ]]; then
```

fukawi2 commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,63 @@
+DB_PATH="/var/db/bedrock.db"
+PARAMS="$PARAMS -db $DB_PATH -nodeHost 0.0.0.0:9000 -serverHost 0.0.0.0:8888"
+
+for i in "${NO_ARG_PARAMS[@]}"
+do
+  :
+  if [ -v "$i" ]; then
```
```suggestion
for i in "${NO_ARG_PARAMS[@]}" ; do
  if [[ -v "$i" ]]; then
```

fukawi2 reviewed (changes requested):
Thanks for the contribution @clifinger! I've made some small suggestions to align with our style guide if that's OK.

While docker isn't something we're using or have a particular interest in right now, this seems well written, neat and self-contained. Having this available to make it easier for new users to get started seems like a good improvement too. I don't see any reason not to merge it.

What do you think @tylerkaraszewski or @coleaeason?

clifinger commented:
Fixed :)

fukawi2 commented:
Thanks! Will wait for some others to have an opportunity to comment before I approve/merge (it's out of hours for them now)

clifinger commented:
Take your time guys, btw I am the CTO of manypixels.co and I will continue to test Bedrockdb little before use it in production for our company ;)

clifinger commented:
I did some tests and some changes to be sure it work:

- Added `bedrock` into `PATH` for the test
- changed `RUN make -j64` to `RUN make -j$(nproc)`

So now I am sure than I can compile and start the cluster with one command
![Screenshot 2024-10-18 122247](https://github.com/user-attachments/assets/7d5f8b23-c211-4ca8-9f4c-24a2d98d2d87)
![Screenshot 2024-10-18 121704](https://github.com/user-attachments/assets/88090c41-95c8-409d-a42a-fca8971981f9)
![Screenshot 2024-10-18 120902](https://github.com/user-attachments/assets/09868c5f-d78f-4797-8db7-143d80c07bcc)

coleaeason commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,70 @@
+      dockerfile: docker/Dockerfile
+    container_name: bedrock-node1
+    ports:
+      - "8887:8888"
```
should this be `8887:8887`?

clifinger replied:
yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

coleaeason commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,70 @@
+      dockerfile: docker/Dockerfile
+    container_name: bedrock-node2
+    ports:
+      - "8889:8888"
```
should this be `8889:8889`?

clifinger replied:
yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

coleaeason commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,70 @@
+    container_name: bedrock-node1
+    ports:
+      - "8887:8888"
+      - "9001:9000"
```
should this be `9000:9000`?

clifinger replied:
yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

coleaeason commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,70 @@
+    container_name: bedrock-node2
+    ports:
+      - "8889:8888"
+      - "9002:9000"
```
should this be `9000:9000`?

clifinger replied:
yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

coleaeason commented on docker/libstuff/bedrock.sh:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
just curious if there is a reason to use this bash script instead of the systemd configs in the repo and systemd to run the process?

clifinger replied:
systemd is never include into docker because of the single process recommandation of docker. If you take a little time to read [https://github.com/phusion/baseimage-docker](https://github.com/phusion/baseimage-docker) you will understand ;)

clifinger replied:
and the second argument is that we need to change a bit the configuration to handle the env. var.:

From the Readme:

## Environment Variables

The following environment variables can be used to configure Bedrock:

- `NODE_NAME`: Name of the Bedrock node
- `PRIORITY`: Priority of the node in the cluster
- `PEER_LIST`: Comma-separated list of peer nodes
- `PLUGINS`: Plugins to enable
- `CACHE_SIZE`: Size of the cache
- `WORKER_THREADS`: Number of worker threads
- `QUERY_LOG`: Query log configuration
- `MAX_JOURNAL_SIZE`: Maximum journal size
- `SYNCHRONOUS`: Synchronous mode setting

tylerkaraszewski commented:
I'm certainly not opposed to adding it.

clifinger commented:
Sorry @coleaeason my answer about the ports concern the 9000+ for peers list, but it's the same meaning map the port from the host to the container. Like that you don't need to change the configuration of files inside docker when you add a node, it' s like unlimited depending of the ressource. that's the power of docker. you just add a node `node3 `with example 8881:8888 and add it to peers list inside the file docker-compose.yml and it scale out of the box.

clifinger commented:
So @coleaeason you can do on the host:

```bash
nc localhost 8887
nc localhost 8888
nc localhost 8889
```
To check if it work ;) it's what I did in my tests in complements of checking the logs output

clifinger commented:
2 months ago, at Manypixels, we decided to change our stack with [Dokploy](https://dokploy.com) and a dedicated server, we did it with no effort and we divided our final bill near by 10 !!!!

clifinger commented:
Now if you want (free and not needed but better), you can make a docker account for Expensify and in your Github action when you merge you just add
```bash
cd docker
docker build -t expensify/bedrockdb:latest -f Dockerfile ..
docker push expensify/bedrockdb:latest
```
Like that we can change the `docker-compose.yml` from:
```yaml
bedrock-node1:
    build:
      context: ..
      dockerfile: docker/Dockerfile
```
to:
```yaml
bedrock-node1:
   image: expensify/expensify/bedrockdb:latest
```
and people will have always the latest build of the tested version ;)

tylerkaraszewski commented:
@coleaeason - I'm ok with this. I will let you resolve any outstanding issues before merging.

clifinger commented:
@coleaeason where are you 😂

coleaeason reviewed (approved):
Thanks for the explanation! Will let @fukawi2 do a final review and merge.

fukawi2 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,95 @@
+
+# Configure ccache
+ENV CCACHE_DIR=/ccache
+RUN mkdir -p $CCACHE_DIR && chmod 777 $CCACHE_DIR
```
NAB: I don't like `777` permissions, but won't block this PR since it's inside a docker container which I presume to be untrusted anyway.

clifinger replied:
yes correct, I don't like also but for docker container it's okay

fukawi2 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,95 @@
+COPY --from=builder /app/bedrock/bedrock /usr/local/bin/
+RUN chmod +x /usr/local/bin/bedrock
+
+RUN mkdir -p /var/db && \
+    touch /var/db/bedrock.db
```
NAB: We use `/var/lib/bedrock/` but again, doesn't really matter inside a container I guess.

clifinger replied:
Same no problem with that as we need to specify it

fukawi2 reviewed (approved):
Thanks again for the contribution! ❤️

fukawi2 commented:
Ah damn, I missed that your commits aren't signed:

![image](https://github.com/user-attachments/assets/e269bc80-2c37-4209-b146-fa49bf0ccfa6)

Are you able to rebase and sign your commits please? You can [use your SSH key](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) if you don't have a GPG key, but for compliance we do need the commits to be cryptographically signed.

clifinger commented:
> Ah damn, I missed that your commits aren't signed:
> 
> ![image](https://private-user-images.githubusercontent.com/139959/379032282-e269bc80-2c37-4209-b146-fa49bf0ccfa6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NTI3NDUsIm5iZiI6MTcyOTY1MjQ0NSwicGF0aCI6Ii8xMzk5NTkvMzc5MDMyMjgyLWUyNjliYzgwLTJjMzctNDIwOS1iMTQ2LWZhNDliZjBjY2ZhNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QwMzAwNDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OGM3MGIyYTBjNzFlNDkwOTY0NjI2OWJmYzM5ZjhhZmViYzFlZDk0ODg4ZmE2MmE5YjM0OWM4MjQwNmM1MmY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YZfIepMAjLcWYHqOxh3WfQJUWUxItT7fgPzWwiqvjgw)
> 
> Are you able to rebase and sign your commits please? You can [use your SSH key](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) if you don't have a GPG key, but for compliance we do need the commits to be cryptographically signed.

Done ;)

clifinger commented:
Not bad, like that my futur commits will be signed

fukawi2 commented:
Thanks! Merged :)
